### DISCUSSION TITLE
5주차 Discussion

### DISCUSSION THREAD

**Discussion by chloe-codes1** [slippStudy/java-network-study#11] [General]
## 📢 책을 읽으면서 궁금한 점들을 자유롭게 적고 대답해 주세요

5주차 진행 방식

> 진행자: 동훈님, 윤혁님
> 

- 9 ~ 10장 읽기
- Discussion 에 질문 or 인상 깊은 점 2개 적어오기
- 4주차 Discussion 이어서 이야기 나누기

**Comment by jinwookh** [comment]
1. 예제 10-3 WebSpiderWithURLConnection에서 실제로 HTTP 통신이 일어나는 코드는 어디일까?(GET 패킷이 전송되는 시점)
2. 왜 URLEncoder는 빈칸을 +로 인코딩하는 것일까?

**Reply by jinwookh** [reply]
구글로 "hello world"를 검색하면, hello+world로 변환이 되어 http 요청을 하는 것 같다.
자동완성 기능에서는 빈칸이 %020%으로 변환이 되는 것 같다.
https://www.google.com/complete/search?q=hello%20wor&cp=9

**Reply by LenKIM** [reply]
실제로 HTTP 통신이 일어나는 코드
![image](https://github.com/slippStudy/java-network-study/assets/22192085/16077a26-fffa-46e0-8569-ea8e59dd0bf9)
handler.connect()

getInputSteam() 호출 할 때 발생하는 것 같다.

**Reply by LenKIM** [reply]
![image](https://github.com/slippStudy/java-network-study/assets/22192085/71c1ea1a-197c-4c11-9c96-896917adff29)

**Reply by chloe-codes1** [reply]
<img width="236" alt="image" src="https://github.com/slippStudy/java-network-study/assets/53922851/b5ffeeff-0bb9-4852-b931-8181344f31e0">


ref. [URL encoding the space character: + or %20?](https://stackoverflow.com/questions/1634271/url-encoding-the-space-character-or-20)

**Reply by LenKIM** [reply]
<img width="668" alt="image" src="https://github.com/slippStudy/java-network-study/assets/22192085/deebd55b-24cd-40ce-912d-2f9b772b08ae">

**Comment by yangyunhyeuk** [comment]
1. 책 내용에 기재된 내용(daum 검색 포탈 내 키워드 입력 시 cgi-bin 디렉토리 접근 부분) 중 cgi-bin 디렉토리 부분이 있는데, 현재는 "https://search.daum.net/search?w=tot&DA=YZR&t__nil_searchbox=btn&sug=&sugo=&sq=&o=&q=%EC%82%AC%EA%B3%BC"와 같은 형태를 띈다.  cgi-bin 디렉토리가 포함된 url의 정체는?

2. http 통신에 있어서 인코딩, 디코딩을 사용하는 이유에는 흔히 아는 텍스트를 변경하는 등의 보안적인 측면 외에 어떠한 것들이 있을까?

**Reply by LenKIM** [reply]
https://www.w3.org/TR/html40/appendix/notes.html#non-ascii-chars

**Comment by LenKIM** [comment]
**인상깊은 내용**
UDP 활용

**질문**
- 책에서는 UDP 를 활용한 예시로 타임서버를 예시로 들고 있다. 우리가 알만한 또다른 예시는 무엇이 있을까?
- UDP 는 유실될 수 있다고 하는데, 어떻게 하면 유실이 될까? 직접 테스트 할 때는 Localhost 라서 유실이 되는 경우가 없는 것으로 보인다. 
- 왜 패킷사이즈가 64K 로 제한되어 있을까?

**Reply by LenKIM** [reply]
URLEncoder 는 무슨 기준으로 '특수한' 문자로 변환시키는 걸까? 아스키 문자란 뭐지? 우리가 사용하는 한글은 아스키 문자가 아니기 때문에 '특수한'문자로 모두다 변환이 된다.

또, 이때 SEO 는 한글도 적용이 될까?

**Reply by LenKIM** [reply]
이론적인 UDP 패킷의 최대 크기는 64KB(65535byte) 이다. 왜냐하면 UDP 헤더에 패킷 크기를 나타내는 부분이 16비트라서 2의 16승으로 나타낼 수 있는 최대값이 한번에 보낼 수 있는 최대 데이터 크기가 되기 때문이다. 그보다 큰 데이터를 보내봤자 그렇게 보낸 줄 알 수가 없다.

**Reply by LenKIM** [reply]
<img width="1163" alt="image" src="https://github.com/slippStudy/java-network-study/assets/22192085/ed18b909-7c64-4aca-9ea4-ab88827bad15">

**Reply by LenKIM** [reply]
https://ko.wikipedia.org/wiki/%EC%8B%A4%EC%8B%9C%EA%B0%84_%EC%A0%84%EC%86%A1_%ED%94%84%EB%A1%9C%ED%86%A0%EC%BD%9C

**Reply by chloe-codes1** [reply]
[Google - 검색엔진 최적화(SEO) 기본 가이드](https://developers.google.com/search/docs/fundamentals/seo-starter-guide?hl=ko)

**Comment by chloe-codes1** [comment]
### 인상깊은 내용

- URLConnection class의 내부를 살펴볼 수 있어 좋았다

### 궁금한 내용

- UDP 통신이 필요한 사례가 책에 있는 내용 외에 어떤 것이 있을지 궁금합니다

**Reply by chloe-codes1** [reply]
UDP 사용 예시

- Live streaming
- Online gaming
- Video chat

**Reply by chloe-codes1** [reply]
[Is DNS TCP or UDP port 53?
](https://www.infoblox.com/dns-security-resource-center/dns-security-faq/is-dns-tcp-or-udp-port-53/#:~:text=DNS%20has%20always%20been%20designed,in%20a%20single%20UDP%20packet.)

> The answer is DNS is mostly UDP Port 53, but as time progresses, DNS will rely on TCP Port 53 more heavily. 
> DNS has always been designed to use both UDP and TCP port 53 from the start1 , with UDP being the default, and fall back to using TCP when it is unable to communicate on UDP, typically when the packet size is too large to push through in a single UDP packet.

**Comment by LenKIM** [comment]
URL / URI 차이?

**Comment by LenKIM** [comment]
URL 이 있는데, URLConnection 은 왜 만든건지?

**Comment by kangbada0728** [comment]
궁금한 내용
- UDP 사례가 HTTP3 부터 기본통신방식으로 바뀌었다는데 어떻게 바뀐걸까

인상깊은 내용
- UDP 에서 보낼 수 있는 패킷의 크기가 그렇게 작을줄은 몰랐다.

**Reply by LenKIM** [reply]
https://www.cloudflare.com/ko-kr/learning/performance/what-is-http3/

**Reply by LenKIM** [reply]
https://www.cloudflare.com/ko-kr/learning/performance/http2-vs-http1.1/

**Reply by chloe-codes1** [reply]
[HTTP/3는 왜 UDP를 선택한 것일까?](https://evan-moon.github.io/2019/10/08/what-is-http3/)
